sql/src/test/java/io/crate/analyze/InsertFromValuesAnalyzerTest.java

```diff
@@ -916,34 +918,65 @@ public void testInsertIntoTableWithNestedPartitionedByColumnAndNullValue() throw
-            "other_id = other_id + 100");
-        assertThat(statement.onDuplicateKeyAssignments().size(), is(1));
+            "other_id = other_id + 100";
+        insertStatements[1] = "insert into users (id, name, other_id) values (1, 'Arthur', 10) " +
+            "on conflict do update set name = substr(excluded.name, 1, 2), " +
+            "other_id = other_id + 100";
+
+        for (String insertStatement : insertStatements) {
+            InsertFromValuesAnalyzedStatement statement = e.analyze(insertStatement);
+            assertThat(statement.onDuplicateKeyAssignments().size(), is(1));
 
-        Symbol[] assignments = statement.onDuplicateKeyAssignments().get(0);
-        assertThat(assignments.length, is(2));
-        assertThat(assignments[0], isLiteral("Ar"));
-        assertThat(assignments[1], isFunction(ArithmeticFunctions.Names.ADD));
-        Function function = (Function) assignments[1];
-        assertThat(function.arguments().get(0), isReference("other_id"));
+            Symbol[] assignments = statement.onDuplicateKeyAssignments().get(0);
+            assertThat(assignments.length, is(2));
+            assertThat(assignments[0], isLiteral("Ar"));
+            assertThat(assignments[1], isFunction(ArithmeticFunctions.Names.ADD));
+            Function function = (Function) assignments[1];
+            assertThat(function.arguments().get(0), isReference("other_id"));
+        }
     }
 
     @Test
     public void testInsertFromValuesWithOnDuplicateKeyInvalidColumnInValues() throws Exception {
         expectedException.expect(ColumnUnknownException.class);
         expectedException.expectMessage("Column does_not_exist unknown");
-       e.analyze("insert into users (id, name) values (1, 'Arthur') " +
+        e.analyze("insert into users (id, name) values (1, 'Arthur') " +
                 "on duplicate key update name = values (does_not_exist)");
     }
 
+    @Test
+    public void testInsertFromValuesWithOnConflictDoUpdateInvalidColumnInExcluded() {
+        expectedException.expect(ColumnUnknownException.class);
+        expectedException.expectMessage("Column does_not_exist unknown");
+        e.analyze("insert into users (id, name) values (1, 'Arthur') " +
+                  "on conflict do update set name = excluded.does_not_exist");
+    }
+
+    @Test
+    public void testInsertFromValuesWithOnConflictDoUpdateAndValueUsage() {
+        expectedException.expect(UnsupportedOperationException.class);
+        expectedException.expectMessage("Can't use VALUES outside ON DUPLICATE KEY");
+        e.analyze("insert into users (id, name) values (1, 'Arthur') " +
+                  "on conflict do update set name = values (name)");
+    }
+
+    @Test
+    public void testInsertFromValuesWithWithDuplicateKeyAndUseOfExcluded() {
+        expectedException.expect(IllegalArgumentException.class);
+        expectedException.expectMessage("Column reference \"excluded.name\" has too many parts");
```

Comment:
Shouldn't we add another if clause like: https://github.com/crate/crate/pull/7087/files#diff-c55569a976310945831be2cae906ec15R104 and have a message ``Can't use EXCLUDED outside ON CONFLICT``? Would seem more consistent to me.

Reply:
In an earlier design, it was like that because the analysis of the two was in the same place. I'd prefer to keep it like it is because we have to add extra logic to detect that and it doesn't add much value. The error message for any other table than "excluded" should be the same in on duplicate key clauses. It is just a precaution here to check that the excluded field provider is not in use.